File: PowerSet.js
// Given a string, print all possible substrings of the given string.
const readline = require('readline');

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

rl.question('Give a string to print all of its substrings: ', (str) => {
    const n = str.length;

    console.log('The Given String is:', str);

    for (let num = 0; num < (1 << n); num++) {
        let sub = '';
        for (let i = 0; i < n; i++) {
            if (num & (1 << i)) {
                sub += str[i];
            }
        }
        console.log(sub);
    }

    rl.close();
});





// const str = prompt("Give a string to print all of its substrings:");
// const n = str.length;

// console.log("The Given String is:", str);

// for (let num = 0; num < (1 << n); num++) {
//     let sub = "";
//     for (let i = 0; i < n; i++) {
//         if (num & (1 << i)) {
//             sub += str[i];
//         }
//     }
//     console.log(sub);
// }

File: PowerSet.py

# //Given a string, print all possible substrings of the given string.

# str=input("Give a string to print all of it's substrings: ")
# n = len(str)
# print ("The Given String is : ", str)
# for num in range (1<<n):
#     sub=""
#     for i in range (n):
#         if(num&(1<<i)):
#             sub+=str[i]
#     print(sub)


    


    # time complexity O(2^n X n)
    # space complexity O(n)


str=input("Give a string to print all of it's substrings: ")
n = len(str)
print ("The Given String is : ", str)
substr=[]
for num in range (1<<n):
    sub=""
    for i in range (n):
        if(num&(1<<i)):
            sub+=str[i]
    substr.append(sub)


for s in substr:
    print(s)

# print("Top Element:", substr[-1])
